cpp/tests/traversal/ms_bfs_test.cu

```diff
@@ -267,7 +267,36 @@ class Tests_MsBfs : public ::testing::TestWithParam<MsBfs_Usecase> {
                                      d_distances.begin(),
                                      d_distances.end(),
                                      static_cast<vertex_t>(0));
-    ASSERT_TRUE(ref_sum > 0);
+
+    auto d_vertex_degree = graph_view.compute_out_degrees(handle);
+
+    thrust::sort(rmm::exec_policy(handle.get_stream()), d_sources.begin(), d_sources.end());
+
+    d_vertices.resize(graph_view.number_of_vertices(), handle.get_stream());
+    
+    thrust::sequence(
+      rmm::exec_policy(handle.get_stream()), d_vertices.begin(), d_vertices.end(), vertex_t{0});
+
+    auto seeds_degree_last = thrust::partition(rmm::exec_policy(handle.get_stream()),
+                                         thrust::make_zip_iterator(d_vertex_degree.begin(), d_vertices.begin()),
+                                         thrust::make_zip_iterator(d_vertex_degree.end(), d_vertices.end()),
+                                         [d_sources = raft::device_span<vertex_t>(d_sources.data(), d_sources.size())] __device__(auto pair_vertex_degree){
+                                            return thrust::binary_search(thrust::seq, d_sources.begin(), d_sources.end(), thrust::get<1>(pair_vertex_degree));
+                                         });
+
+    auto seeds_degree_size = thrust::distance(thrust::make_zip_iterator(d_vertex_degree.begin(), d_vertices.begin()), seeds_degree_last);
+    
+    d_vertex_degree.resize(seeds_degree_size, handle.get_stream());
+
+    auto seeds_degree_sum = thrust::reduce(rmm::exec_policy(handle.get_stream()),
+                                           d_vertex_degree.begin(),
+                                           d_vertex_degree.end(),
+                                           static_cast<vertex_t>(0),
+                                           thrust::plus<vertex_t>());
```

Comment:
Are you just summing the degree of all the sources?
```
auto seed_degree_first = thrust::make_transform_iterator(d_sources.begin(), cuda::proclaim_return_type<edge_t>([degrees = raft::devie_span<edge_t const>(d_vertex_degree.data(), d_vertex_degree.size())]__device__(auto v) { return degrees[v]; }));
auto seed_degree_sum = thrust::reduce(handle.get_thrust_policy(), seed_degree_first, seed_degree_first + d_sources.size());
```
will be sufficient.

Reply:
And if you have time, could you update this test to use a same mechanism as other C++ tests?

We don't use
```
    hr_timer.start("msbfs");
    cudaProfilerStart();
``` (performance measurement should be enabled with --perf command line option).

We check with edge masking and we take check_correctness flag. We have rmat_small tests for correctness check and rmat benchmark tests without correctness check and so on.

This test is written in its own way... making our codebase less consistent and harder to maintain.

Reply:
Got it. Let me refactor this

Reply:
> Are you just summing the degree of all the sources?
Yep. Ok I will use `thrust::make_transform_iterator` as you suggested

Reply:
Actually I am summing the degree of all seeds, not sources

Reply:
> And if you have time, could you update this test to use a same mechanism as other C++ tests?

Wouldn't this break the current structure of the test? In fact, this test relies on creating multiple disjoint graphs while selecting one seed per graph which is possible by generating multiple edgelist. I am not sure how we can adapt this with `file_test` and `rmat_small_test` in a way that matches the other C++ tests. Do I just select random seeds from one edgelist?

Reply:
You could mimic the structure of the other tests but only use RMAT.  Perhaps a comment as to why the file tests are omitted.

Within the test you could still use the same "create a series of RMAT graphs" approach.  There are several tests that don't just create a graph but create the edge list, do some things and then create the graph from the resulting edge list.

But if the overall structure was more consistent that would be beneficial for maintenance.

Reply:
> Perhaps a comment as to why the file tests are omitted.

With the approach you suggested which relies on finding the connected components of a graph and selecting the MsBFS seeds from the representative of each component (min vertex IDs), we no longer need to omit the file tests